From: Michal Kubecek <mkubecek@suse.cz>
Date: Thu, 20 Dec 2012 19:38:18 +0100
Subject: xfrm6: count extension headers into payload length
Patch-mainline: Not yet, upstream needs additional fix
References: bnc#794513

In ESP transport mode, if decrypted payload contains IPv6
extension headers (e.g. fragmentation if it is sent using a
route with allfrag flag), we need data point to first extension
header rather than transport header when leaving
xfrm6_transport_input(). Otherwise the packet is classified as
invalid by netfilter conntrack.

Signed-off-by: Michal Kubecek <mkubecek@suse.cz>
---
 net/ipv6/xfrm6_mode_transport.c |    5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/net/ipv6/xfrm6_mode_transport.c b/net/ipv6/xfrm6_mode_transport.c
index 4e34410..80be103 100644
--- a/net/ipv6/xfrm6_mode_transport.c
+++ b/net/ipv6/xfrm6_mode_transport.c
@@ -53,9 +53,10 @@ static int xfrm6_transport_input(struct xfrm_state *x, struct sk_buff *skb)
 			skb_network_header(skb), ihl);
 		skb->network_header = skb->transport_header;
 	}
-	ipv6_hdr(skb)->payload_len = htons(skb->len + ihl -
-					   sizeof(struct ipv6hdr));
 	skb_reset_transport_header(skb);
+	__skb_push(skb, ihl - sizeof(struct ipv6hdr));
+	ipv6_hdr(skb)->payload_len = htons(skb->len);
+
 	return 0;
 }
 
-- 
1.7.10.4

